Add service to glossary (#2111)
open-telemetry/opentelemetry-specification: specification/glossary.md

creating a new `Tracer` or `Meter` (see [Obtaining a Tracer](trace/api.md#tracerprovider)/[Obtaining a Meter](metrics/api.md#meterprovider)).
The name/version pair identifies the [Instrumentation Library](#instrumentation-library).

> ### Service

Reviewer: is this the only attempt to define the service in the spec, or do we have other places?

Author: I didn't find any other place where service is defined in an explicit manner. The one place talking about service the most is the semantic conventions:

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/README.md#service

Although there is no wording on what a service is (or what it is not)

I went through all pages of the search for `service` and didn't find any definition:

https://github.com/open-telemetry/opentelemetry-specification/search?q=service

> `Services` expose unique functions to another and are identified by a namespace and a name. There are one or more instances of the same service called `Service Instance`. Often such a service instance is enclosed within an operating system process. There could be multiple versions of the same service.

Reviewer 2: I don't think a service needs to have an unique function. And I think it exposes it's functionality potentially also to end users, or it may not expose any functionality (client-only service)

Author: Thanks for the feedback, how about:

> `Services` expose (certain, specific, <blank>) functions to another and to end users

Reviewer 2: I wonder if that is the correct definition for OTel. Can a process that just calls other services (e.g. a rich client) still be a service? I would think so.

Reviewer: I think [Martin Fowler's definition][1] works well for us

> ... a single **application** as a suite of small **services**, each running in its own process and communicating with lightweight mechanisms, often an HTTP resource API. These services are built around business capabilities and independently deployable by fully automated deployment machinery. There is a bare minimum of centralized management of these services, which may be written in different programming languages and use different data storage technologies.

I would replace
```diff
- communicating with lightweight mechanisms, often an HTTP resource API
+ communicating over the network
```

And I think it's good to introduce the notion of an "application" or "distributed application like Gmail or Instagram"

[1]: https://martinfowler.com/articles/microservices.html

Author: So, if I extract from that a definition it would be:

> An application is a suite of services, each running in its own process and communicating with each other

I like that definition, it's also close to the Glossary of the website (https://opentelemetry.io/docs/concepts/glossary/)

* Is a service always running in its own (operating system) process? -- I remember that there where some java webservers that allow you to put multiple logical apps into one runtime. Not that anyone recommends doing that, but it is possible.
* Communication over the network: if two processes run on the same host, they could use different mechanisms to communicate (shared memory, filesystem, ...), is this something OTel would cover?

This would only give a definition for "Application", would it then make sense to put in a definition for "Service" and put a "See definition for Application" link there?

Reviewer: We can also borrow a bit from SOA.

A **service** is a component of a larger distributed **application**. A service logically represents a repeatable business activity with a specified outcome. Each service usually runs in its own process and exchanges information with other services via some form of inter-process communication mechanism, such as network or shared storage.

Author: that's a good one! Trying to slice it:

**Application**
An application is a suite of one or more services.

**Service**
A `service` is a logical representation of a repeatable activity with a specified outcome. There are one or more `service instances`, usually running in their own operating system processes, communicating with each other.

Reviewer: The paragraph was my editing of how I can see it defined in OTEL. I don't think we need an explicit entry for the application. Your definition has an issue where it looks like service instances are inter-communicating, instead of different services.

Author: > The paragraph was my editing of how I can see it defined in OTEL.
Sorry, it was late for me already ;-)

> I don't think we need an explicit entry for the application.
Ok, makes sense.

> Your definition has an issue where it looks like service instances are inter-communicating, instead of different services.
I am not sure if I understand: my assumption was that a service is a grouping of one or many service instances: service instances come and go (in time) or are scaled up and down, but the service always stays the same, so I thought also the service instance is the thing running in an OS process and doing the communication.

Reviewer: yes, I agree with everything you just said, I meant that the actual wording implied instances talk to each other, which isn't the property we are concerned with. A "service" is a logical construct, and services talk to each other. Service instance is a physical manifestation of a service (which is where it gets the service name from). If we more clearly separate these concepts then we'd be ok.

Author: Gotcha! So using your words and adding some words from the website glossary around HA/scalability:

A **service** is a component of a larger distributed **application**. A service logically represents a repeatable business activity with a specified outcome. A service exchanges information with other services via some form of inter-process communication mechanism, such as network or shared storage.

A **service instance** is a physical manifestation of a service, usually running in its own operating system process. Multiple service instances are typically deployed at the same time for high availability and scalability.

Reviewer: +1

Author: Updated the definition to that wording via the last commit.

creating a new `Tracer` or `Meter` (see [Obtaining a Tracer](trace/api.md#tracerprovider)/[Obtaining a Meter](metrics/api.md#meterprovider)).
The name/version pair identifies the [Instrumentation Library](#instrumentation-library).

### Service

> A `service` is a component of a larger distributed `application`. A service logically represents a repeatable business activity with a specified outcome. A service exchanges information with other services via some form of inter-process communication mechanism, such as network or shared storage.

Reviewer 3: If we feel that `service` is not self-explanatory, then we cannot define it via `application`, which itself is not defined.

Author: That's a good point. A proposal how service defines the word `application`:

> ~~A `service` is a component of a larger distributed `application`.~~ A service logically represents a repeatable business activity with a specified outcome. A service exchanges information with other services via some form of inter-process communication mechanism, such as network or shared storage. A group of services are forming an `application`.

Note: we have to pay attention with the word "application" since it is heavily overloaded, see the conversation around service vs frontend app(lication). `Distributed Application` would make this more clear, but as per your other comment, this would not make sense for an application only having a single service.

Reviewer 4: Let's not correlate at all with "application" to avoid confusion as you suggested:

```
A service logically represents a repeatable business activity with a specified outcome. A service exchanges information with other services via some form of inter-process communication mechanism, such as network or shared storage.
```

Reviewer 5: I don't understand what "logically represents a repeatable business activity with a specified outcome" means. That doesn't sound like a service to me. Also, why does a service need to exchange information with other services? That seems like something a service might do, but it's certainly not a requirement.

Author: > "logically represents a repeatable business activity with a specified outcome"
See https://github.com/open-telemetry/opentelemetry-specification/pull/2111#discussion_r746741718: this comes from the service definition of SOA. What alternative would you suggest?

> Also, why does a service need to exchange information with other services? That seems like something a service might do, but it's certainly not a requirement.
Would adding a "may", i.e. "A service may exchange..." help?

Reviewer 5: Higher level answer provided as a comment on the PR itself, for broader visibility.

> A `service` is a component of a larger distributed `application`. A service logically represents a repeatable business activity with a specified outcome. A service exchanges information with other services via some form of inter-process communication mechanism, such as network or shared storage.

Reviewer 3: Is `distributed` important here? If my application is not distributed, then I don't have any services by this definition?

Author: See my note above, if we remove distributed it can be confused much more easily with a frontend application. Maybe there is a better term.

Reviewer 4: See my note, let's remove any reference to "application" for the moment.

> A `service instance` is a physical manifestation of a service, usually running in its own operating system process. Multiple service instances are typically deployed at the same time for high availability and scalability.

Reviewer 3: ```suggestion
A `service instance` is a physical manifestation of a service, usually running in its own operating system process. Multiple service instances may be deployed at the same time for high availability and scalability.
```
